Improve handling of unconnecting headers

sdaftuar opened the pull request:
I found a situation on testnet where a peer could send a header that wouldn't be accepted locally, because of the 2-hour-in-the-future rule.  This would then cause BIP-130 headers sync to fail, because the sender doesn't realize that the recipient didn't accept the header, and so continues to send headers (one-by-one) that build on top of the rejected header, until eventually the sender is disconnected.

This PR tries to improve upon the situation: if the first header in a received headers message doesn't connect, try once to request connecting headers.

If we get two headers messages in a row that don't connect, give up on trying to request connecting headers (to avoid infinite looping with the remote peer), until our peer delivers a headers message that does connect.

If this approach is acceptable, I think we should backport to 0.12 (looks like it doesn't merge cleanly, but I can open a new PR for the backport).

cc: @TheBlueMatt @sipa

laanwj commented:
Needs rebase after #8306

sdaftuar commented:
Rebased.

mrbandrews commented on src/main.cpp:
```diff
@@ -5773,6 +5776,20 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+            if (nodestate->fLastHeadersConnected) {
+                nodestate->fLastHeadersConnected = false;
+                pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexBestHeader), uint256());
+                return true;
```
Print a log message here?

pstratem commented:
@sdaftuar can you split this into two commits, one with changes and a second with tests?

sdaftuar commented:
Added logging, and split into two commits.

sdaftuar committed:
Improve handling of unconnecting headers

When processing a headers message that looks like a block announcement,
send peer a getheaders if the headers message won't connect.

Apply DoS points after too many consecutive unconnecting headers messages.

sdaftuar committed:
Add test for handling of unconnecting headers

pstratem commented:
@sdaftuar thank you

mrbandrews commented:
ACK

sipa commented:
I've tried to reason through what would happen if two subsequent blocks happen with a very small time between them, before the peer has time to respond to the getheaders sent as response to the first block, but couldn't find any problem. I believe that convergence would only fail in case every future pair of blocks happens with a sufficiently short time between them.

sipa commented:
ACK

theuni commented:
Is it necessary to recover from this more than once per remote-node? As-is, it looks like someone could just respond with alternating connecting/non-connecting headers forever since a ban score isn't set. I can't imagine any motivation for doing so, though...

Also, I don't think this should be allowed during IBD? Seems the approach here could be unified with the one when a non-connecting CMPCTBLOCK is received.

sdaftuar commented:
Discussed with @gmaxwell on IRC yesterday.  He pointed out one drawback to this approach is that if we were to receive 2 blocks in short succession, such that the first block was still not valid according to the 2hr rule when the second one was received, then we would still fall out of sync.

So I'm going to push an alternate approach, where instead of just trying once to recover, we do the following: when receiving a HEADERS message with at most a few headers, if the first header doesn't connect, increment a counter and send the peer a getheaders.  If the counter gets too big, assign DoS points, but if a connecting header comes in before that happens, reset the counter to 0.

@theuni Since DoS points don't ever decay back to 0, I was reluctant to assign DoS points at all when this type of thing happens, because otherwise if you were up and running long enough, you'd eventually disconnect your peers as these situations accumulated points over time.

One difference between this and the CMPCTBLOCK handling is that we have to worry about potential infinite looping if our peer is broken, since we're sending a getheaders in response to a headers message.

sdaftuar commented:
Updated with new approach.  @gmaxwell How does this look?

gmaxwell commented:
This looks reasonable to me. I am testing it.

sipa commented:
re-utACk

gmaxwell commented:
ACK

laanwj commented:
utACK e91cf4b

rebroad commented on src/main.cpp:
```diff
@@ -5773,6 +5776,35 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             return true;
         }
 
+        CNodeState *nodestate = State(pfrom->GetId());
```
may I ask why you use a function GetId() when you could just use the variable id? i.e. pfrom->id instead of pfrom->GetId

sipa replied:
Can you please stop ask basic programming questions? You can learn these things online, or on various other forums (including stackoverflow or chat channels) that don't require the developers of the project to waste their time teaching you.

You're free to ask for rationale behind design decisions, but this is getting unreasonable.

One last time: it's generally good style to use accessors instead of field variables directly. No, we don't do that everywhere yet. No, that doesn't mean we should continue that practice.
